src/json.c

```diff
@@ -7,41 +7,77 @@
 #include <string.h>
 
 // REJSON APIs
-RedisJSONAPI_V1 *japi = NULL;
+RedisJSONAPI *japi = NULL;
+int japi_ver = 0;
 
 ///////////////////////////////////////////////////////////////////////////////////////////////
 
 void ModuleChangeHandler(struct RedisModuleCtx *ctx, RedisModuleEvent e, uint64_t sub,
                          RedisModuleModuleChange *ei) {
   REDISMODULE_NOT_USED(e);
-  if (sub != REDISMODULE_SUBEVENT_MODULE_LOADED || strcmp(ei->module_name, "ReJSON") || japi)
+  if (sub != REDISMODULE_SUBEVENT_MODULE_LOADED || japi || strcmp(ei->module_name, "ReJSON"))
     return;
   // If RedisJSON module is loaded after RediSearch need to get the API exported by RedisJSON
 
   if (!GetJSONAPIs(ctx, 0)) {
     RedisModule_Log(ctx, "error", "Detected RedisJSON: failed to acquire ReJSON API");
-  }
-  //TODO: Once registered we can unsubscribe from ServerEvent RedisModuleEvent_ModuleChange
-  // Unless we want to handle ReJSON module unload
+  }  
 }
 
 //---------------------------------------------------------------------------------------------
 
 int GetJSONAPIs(RedisModuleCtx *ctx, int subscribeToModuleChange) {
-    japi = RedisModule_GetSharedAPI(ctx, "RedisJSON_V1");
+    japi = RedisModule_GetSharedAPI(ctx, "RedisJSON_V2");
     if (japi) {
+      japi_ver = 2;
+      RedisModule_Log(ctx, "notice", "Acquired RedisJSON_V2 API");
+    } else {
+      japi = RedisModule_GetSharedAPI(ctx, "RedisJSON_V1");
+      if (japi) {
+        japi_ver = 1;
         RedisModule_Log(ctx, "notice", "Acquired RedisJSON_V1 API");
-        return 1;
+      }
+    }
+    if (japi) {
+      return 1;
     }
     if (subscribeToModuleChange) {
-        RedisModule_SubscribeToServerEvent(ctx, RedisModuleEvent_ModuleChange,
-                                           (RedisModuleEventCallback) ModuleChangeHandler);
+      RedisModule_SubscribeToServerEvent(ctx, RedisModuleEvent_ModuleChange,
+                                         (RedisModuleEventCallback) ModuleChangeHandler);
     }
     return 0;
 }
 
 ///////////////////////////////////////////////////////////////////////////////////////////////
 
+JSONPath pathParse(const char *path, RedisModuleString **err_msg) {
+  if (japi_ver >= 2) {
+    return japi->pathParse(path, RSDummyContext, err_msg);
+  } else {
+    *err_msg = NULL;
```

Comment:
Why not setting an error message here?

Reply:
Error message is given by the out param `err_msg`, set by the callee

Reply:
Yes but here you do not call the callee so maybe something like `json path parser are only supported on v2 api`?

Reply:
Returning NULL and a NULL `err_msg` means "API is not available" - I want to be able to differentiate between "API is available and parsing failed", which returns NULL and a non-NULL `err_msg`.